### ISSUE TITLE
Passenger session is closed when client drops, but app may still be busy

### ISSUE THREAD

**Issue by FooBarWidget** [phusion/passenger#521]
Clarification edit:
If the client disconnects (Broken pipe), Passenger releases the associated session and resources, irrespective of what the app might still be doing.

_From [dmitry.andrianov](https://code.google.com/u/101102486751573777399/) on November 21, 2009 00:13:35_

What steps will reproduce the problem? 1. create a "page" which sends something into output every 15 seconds and
keeps running for a long time (or forever).
2. direct your browser to that page
3. passenger-status will indicate there is one session in one of your
server processes.
4. hit browser's Stop button.
5. in approximately 20-30 seconds (two writes from server to client) Apache
will detect connection is closed and Passenger will put into the log:

  Either the vistor clicked on the 'Stop' button in the web browser, or the
visitor's connection has stalled and couldn't receive the data that Apache
is sending to it. As a result, you will probably see a 'Broken Pipe' error
in this log file. Please ignore it, this is normal. You might also want to
increase Apache's TimeOut configuration option if you experience this
problem often.

6. Now passenger-status shows 0 active sessions for the server process.

However the server process is still busy with your request! It is not idle!
Only on the next attempt to write something back it will receive EPIPE and
finish processing of current request. But depending on the nature of the
page this can happen in some distant time.

The problem is that since passenger considers the server process inactive -
it can direct next user request to it. And this request can spend a lot of
time waiting for completion of previous request even though there are lots
of other servers idle.

CentOS, passenger 2.2.7, Apache2

The correct behavior would be not considering the process inactive until it
really finishes processing of the request.

_Original issue: http://code.google.com/p/phusion-passenger/issues/detail?id=421_

**Comment by OnixGH** [comment]
In fact, noone will receive the content because Passenger will not accept it anymore from the application (the EPIPE the original poster mentioned).

Closing this issue because I don't see how we could change this mechanism. The only way the application knows the client is gone is because Passenger drops the Passenger/Application connection (by stopping the session), and once that is done Passenger has no way of knowing whether the app is still busy.

In other words, if Passenger wouldn't stop the session the App would remain active forever without knowing the client is long gone.
